vllm/model_executor/models/ultravox.py

```diff
@@ -423,9 +487,26 @@ def forward(self, input_ids: torch.Tensor, positions: torch.Tensor,
             inputs_embeds = self.language_model.model.get_input_embeddings(
                 input_ids)
 
-            inputs_embeds = merge_multimodal_embeddings(
-                input_ids, inputs_embeds, audio_embeddings,
-                _AUDIO_PLACEHOLDER_TOKEN)
+            prefill_metadata = attn_metadata.prefill_metadata
+            if not prefill_metadata:
+                raise ValueError(
+                    "Audio inputs should only be handled during prefill.")
+
+            seq_lens_tensor = getattr(prefill_metadata, "seq_lens_tensor",
+                                      None)
+            if seq_lens_tensor is None:
+                raise ValueError(
+                    "AttentionMetadata should include a seq_lens_tensor.")
+
+            seq_start_loc = getattr(prefill_metadata, "seq_start_loc", None)
+            if seq_start_loc is None:
+                raise ValueError(
+                    "AttentionMetadata should include seq_start_loc.")
+
+            inputs_embeds = merge_partial_multimodal_embeddings(
+                inputs_embeds, positions, seq_start_loc, seq_lens_tensor,
+                audio_embeddings, audio_input["placeholder_offsets"],
+                audio_input["placeholder_lengths"])
```

Comment:
To avoid duplicating the above checks across different models, let's pass `attn_metadata` to `merge_partial_multimodal_embeddings` and perform the check inside that function.

Reply:
I also wonder whether we can parse `placeholder_offsets` and `placeholder_lengths` in a model-agnostic manner, since it looks like you're just forwarding the kwargs passed to the model (only change being removing the `audio_` prefix) anyway.

Reply:
I pushed changes for the former -- I'm certainly open to the latter but could use some additional guidance on that. I added the `audio_`/`image_` prefixes to make the plumbing more explicitly on the model side of things with an eye towards models that support multiple non-text modalities, but it seems like it could be doable in a central place as well.

Reply:
I would check out where the attention metadata is constructed and see if we can access the placeholder offsets and lengths at that stage.

Reply:
Sorry for the delay -- I'm struggling a bit to navigate the various layers here. Is something like this along the lines of what you had in mind?

```python
class AttentionMetadata:
    ...
    mm_placeholder_mask: Optional[Dict[str, torch.Tensor]] = None
    mm_data_mask: Optional[Dict[str, torch.Tensor]] = None
```

```python
class InterDataForSeqGroup:
    ...
    mm_placeholder_mask: Optional[Dict[str, List[int]]] = None,
    mm_data_mask: Optional[Dict[str, List[List[int]]]] = None,
```

such that the merge in the model (modulo handling `NestedTensors` etc.) is roughly

```python
inputs_embeds[attn_metadata.mm_placeholder_mask["audio"]] = audio_embeddings[attn_metadata.mm_data_mask["audio"]]
```

?

Reply:
Instead in attention metadata (top layer), I am thinking of putting this in the inner prefill metadata.

Reply:
It's the same type for all three unfortunately:

```
class AttentionMetadata:
    ...
    @property
    @abstractmethod
    def prefill_metadata(self) -> Optional["AttentionMetadata"]:
        """Return the attention metadata that's required to run prefill
        attention."""
        pass

    @property
    @abstractmethod
    def decode_metadata(self) -> Optional["AttentionMetadata"]:
        """Return the attention metadata that's required to run decode
        attention."""
        pass
```

but I can null it out for decode metadata in any event. Ok, will take a stab at this and try to have something up for review later today.

Reply:
Alright, I reworked the PR (significantly) -- PTAL :)

An additional benefit of this approach is that we now easily (/centrally) filter the MM data to only the items that intersect the current prefill range.